lua/sh-tas-utils/betterchatprint.lua

```diff
@@ -1,48 +1,107 @@
+local ARGS_LENGTH_BITS = 8
+local BOOLEAN_BITS = 1
+local COLOUR_CHANNEL_BITS = 8
+local NULL_TERMINATOR_BITS = 8
+
+local MAX_MESSAGE_SIZE_BITS = 4 * 1024 * 8
+local MAX_ARGS = 2 ^ ARGS_LENGTH_BITS - 1
+
 if SERVER then
 	util.AddNetworkString("TASUtils.ChatPrint")
 
-	local function encodeMsg(args)
-		for _, v in ipairs(args) do
+	--- Returns the size of a chat print net message in bits
+	---@param args any[]
+	---@return integer
+	function TASUtils.getChatPrintMessageSizeBits(args)
+		local size = ARGS_LENGTH_BITS
+
+		for i, v in ipairs(args) do
+			if i > MAX_ARGS then
+				break
+			end
+
+			size = size + BOOLEAN_BITS
+
+			if IsColor(v) then
+				size = size + COLOUR_CHANNEL_BITS * 3
+			else
+				size = size + #tostring(v) * 8 + NULL_TERMINATOR_BITS
+			end
+		end
+
+		return size
+	end
+
+	local function writeMessage(args)
+		net.WriteUInt(#args, ARGS_LENGTH_BITS)
+
+		for i, v in ipairs(args) do
+			if i > MAX_ARGS then
+				break
+			end
+
 			local isColor = IsColor(v)
 			net.WriteBool(isColor)
 
 			if isColor then
-				net.WriteColor(v, false)
+				net.WriteUInt(v.r, COLOUR_CHANNEL_BITS)
```

Comment:
This is functionally identical to `net.WriteColor`, any particular reason you wanted to change it this way?

Reply:
I wanted to increase the confidence that the size calculation function and this writing function are always 1:1. By doing this we avoid any possible regressions due to changes to the WriteColor util (however unlikely)

Also, the WriteColor function has some checks that are redundant in this case like asserting that the value is indeed a colour (which we’re already doing above)

Reply:
Also just checked again it it *unpacks the colour object*, which is even worse: https://github.com/Facepunch/garrysmod/blob/master/garrysmod/lua/includes/extensions/net.lua#L103-L116

i’m inclined to avoid the function in general now too